### ISSUE TITLE
Multiprocess cache dataset CPU memory usage

### ISSUE THREAD

**Issue by wyli** [Project-MONAI/MONAI#5930]
Although the cached list of items of cachedatset can be shared among multiple processes, the memory usage can be unecessarily large as described in https://github.com/pytorch/pytorch/issues/13246

Would be great to review the issue and provide a reasonable default solution to avoid this. https://github.com/pytorch/pytorch/issues/13246#issuecomment-905703662

It seems our recent runtime cache can workaround it because the multiprocess list removes cpython refs https://github.com/Project-MONAI/MONAI/blob/0d38e4b1522b6457e17a89139cd116aa4bc5192f/monai/data/dataset.py#L848

  Cc @Nic-Ma @ericspod @rijobro @yiheng-wang-nv

**Comment by wyli** [comment]
Possible related issues
- https://github.com/Project-MONAI/MONAI/issues/3116
- https://github.com/Project-MONAI/MONAI/discussions/5850

**Comment by Nic-Ma** [comment]
@yiheng-wang-nv is trying to verifying the issue in `CacheDataset` with `runtime_cache = False / True`.

Thanks.

**Comment by yiheng-wang-nv** [comment]
Update:
Our `CacheDataset` has the same issue no matter which value is used for `runtime_cache`
Detailed experiment results can be found in this sheet:
https://docs.google.com/spreadsheets/d/15h9vbWauBEWgX2faNYCm0PM7ARM6wUc8u9i1GBLPSyc/edit?usp=sharing

Now I'm working on checking the solutions mentioned above.

**Comment by yiheng-wang-nv** [comment]
Update 2:

After discuss with @wyli , I added two more kinds of tests (figures are added in the sheet mentioned above):
1. produce randomly generated 3d data as the input, and a MONAI transform: `LoadImageD` 
2. add one more random transform
All of the experiments show the tendency that:
1. when `num_workers` is 0, there is almost no memory fluctuation in each epoch
2. when `num_workers > 0`, in each epoch, more memory is used with the increasing number of iterations.
3. when `num_workers > 0`, larger `num_workers` require more memory in total.
4. If running for a long time, in some epochs, the memory used is extraordinarily higher than other epochs.

Counting memory used cannot get deeper analysis, thus I will add some experiments to count both shared memory and "private" memory used on each process (as shown in the following figure), and then using methods suggested by others in the reference tickets to see if any of them are helpful. I will try to finish the analysis within one day.

![image](https://user-images.githubusercontent.com/68361391/217561954-63f26bb9-9d50-46d5-967b-010ad3efc52e.png)

**Comment by yiheng-wang-nv** [comment]
Update 3:

I calculated memory from dataloader related processes only, and count both shared memory and PSS (private memory + shared memory), and tendency 4 (mentioned in update 2) does not exist.
In addition, we can see that shared memory (per process) is stable no matter how many processes we have.
The relationship between PSS and `num_workers` meets tendency 3.

To sum up:

1. cachedataset cannot workaround the existing pytorch dataset memory issue, no matter which value is used on `runtime_cache`.
2. I did not detect other issues on cachedataset from all of the experiments, and it performs similar as pytorch dataset and monai dataset.
3. there are many variables if need to analyze the memory, such as random transforms, num of workers, OS, input size, batch size and so on. Tricks from the reference is not a rule of thumb and may not work (or need to make trade-off with other parts, such as speed).

All experiments are running on Unix OS, and my pc has 32GB memory. The code of this update is attached.

Hi @wyli @Nic-Ma @ericspod @rijobro , do you have any suggestions on this ticket? Welcome to discuss.
[cachedataset_monai_component_detailed_check.ipynb.zip](https://github.com/Project-MONAI/MONAI/files/10698759/cachedataset_monai_component_detailed_check.ipynb.zip)

**Comment by wyli** [comment]
thanks Yiheng, the tests make sense, could you please add the code to one of the existing integration tests? for example  https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_integration_segmentation_3d.py which is a typical segmentation pipeline.

**Comment by yiheng-wang-nv** [comment]
> https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_integration_segmentation_3d.py

Sure. Which test is needed? Check if memory used is increased (a lot) after multiple epochs?

**Comment by wyli** [comment]
I think we can just introduce the memory monitoring to the test, without changing the actual training components. The goal is to check by default whether the memory usage behaviour is consistent across different versions and ci environment.

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Thanks for the detailed experiments and summary.
I checked your experiment results, seems when `runtime_cache="process"` the memory usage didn't increase when increasing the `num_workers`?
<img width="1135" alt="image" src="https://user-images.githubusercontent.com/27533405/218025195-1d3ad504-ad06-4111-84e0-7255f30d6924.png">
Could you please help me understand it?

Thanks in advance.

**Comment by yiheng-wang-nv** [comment]
> Hi @yiheng-wang-nv ,
> 
> Thanks for the detailed experiments and summary. I checked your experiment results, seems when `runtime_cache="process"` the memory usage didn't increase when increasing the `num_workers`? <img alt="image" width="1135" src="https://user-images.githubusercontent.com/27533405/218025195-1d3ad504-ad06-4111-84e0-7255f30d6924.png"> Could you please help me understand it?
> 
> Thanks in advance.

Update: I've discussed with @Nic-Ma offline, the vertical axis is different for each figure, and the tendency in "processes" mode is actually close to other modes.
